File: src/tokenizer.cpp
//
// Created by Sunever Liu on 2015/11/21.
//

#include "tokenizer.h"

ifstream input;


bool init_tokenizer(string filename) {
    try {
        input.open(filename, ios::in);
        if (!input)
            throw FileOpenError();
    } catch(FileOpenError e) {
        cerr << e << endl;
        return;
    }

    return true;
}

Token get_token() {
    string tokenBuffer = "";
    string character = "";
    Token token = {ERRTOKEN, "", 0.0, NULL};
    token.lexeme = tokenBuffer;

    input.read(character, 1);

    while(character[0] < 30) {
        input.read(character, 1);
    }

    tokenBuffer = tokenBuffer + character;

    input >> tokenBuffer;

    if (isalpha(character[0])) {
        input.read(character, 1);
        while(isalpha(character[0])){
            input.read(character, 1);
            tokenBuffer = tokenBuffer + character;
        }
        for(Token &t : TokenTab) {
            if (t.lexeme == tokenBuffer)
                return t;
        }

    }
    else if(isdigit(tokenBuffer[0])) {
        input.read(character, 1);
        while(isalpha(character[0])){
            input.read(character, 1);
            tokenBuffer = tokenBuffer + character;
        }
        token.type = CONST_ID;
        token.lexeme = tokenBuffer;
        istringstream iss(tokenBuffer);
        iss >> token.value;

        return token;
    }
    else {
        switch(character[0]) {
            case '*':
                input.read(character, 1);
                if(character[0] != '*') {
                    token.type = MUL;
                    token.lexeme = "*";
                }
                else {
                    token.type = POWER;
                    token.lexeme = "**";
                }
                return token;
            case '/':
                input.read(character, 1);
                if(character[0] != '/') {
                    token.type = DIV;
                    token.lexeme = "/";
                }
                else input.getline();
                return token;
            case '-':
                input.read(character, 1);
                if(character[0] != '-') {
                    token.type = MINUS;
                    token.lexeme = "-";
                }
                else input.getline();
                return token;
            case '+':
                token.type = PLUS;
                token.lexeme = "+";
                return token;
            case ',':
                token.type = COMMA;
                token.lexeme = ",";
                return token;
            case ';':
                token.type = SEMICO;
                token.lexeme = ";";
                return token;
            case '(':
                token.type = L_BRACKET;
                token.lexeme = "(";
                return token;
            case ')':
                token.type = R_BRACKET;
                token.lexeme = ")";
                return token;

        }
    }

    return token;
}

void close_tokenizer() {
    input.close();
}

File: tokenizer.py
__author__ = 'Sunever Liu'

from init import tokenTable

def get_token(input, currentChar):
    tokenBuffer = ""

    char = currentChar
    if char == ' ':
        char = input.read(1)

        while(char < 40 or char == '\n'):
            char = input.read(1)

    tokenBuffer += char

    if char.isalpha():
        char = input.read(1)

        while (char.isalpha()):
            tokenBuffer = tokenBuffer + char
            char = input.read(1)


        for item in tokenTable:
            if tokenBuffer == item[1]:
                return item, char

    elif char.isdigit():
        char = input.read(1)

        while (char.isdigit()):
            tokenBuffer = tokenBuffer + char
            char = input.read(1)

        return ('CONST_ID', tokenBuffer, float(tokenBuffer), 'NULL'), char

    elif char == '*':
        char = input.read(1)
        if(char == '*'):
            return ('POW', '**', 0.0, 'NULL'), char
        else:
            return ('MUL', '*', 0.0, 'NULL'), char

    elif char == '/':
        char = input.read(1)
        if(char == '/'):
            input.readline()
            return ('ENTER', '', 0.0, 'NULL'), char
        else:
            return ('DIV', '/', 0.0, 'NULL'), char

    elif char == '-':
        char = input.read(1)
        if(char == '-'):
            input.readline()
            ('ENTER', '', 0.0, 'NULL'), char
        else:
            return ('MINUS', '-', 0.0, 'NULL'), char

    elif char == '+':
        char = input.read(1)
        return ('PLUS', '+', 0.0, 'NULL'), char
    elif char == ',':
        char = input.read(1)
        return ('COMMA', ',', 0.0, 'NULL'), char
    elif char == ';':
        char = input.read(1)
        return ('SEMICO', ';', 0.0, 'NULL'), char
    elif char == '(':
        char = input.read(1)
        return ('L_BRACKET', '(', 0.0, 'NULL'), char
    elif char == ')':
        char = input.read(1)
        return ('R_BRACKET', ')', 0.0, 'NULL'), char
    elif char == '\n':
        char = input.read(1)
        return ('ENTER', 'ENTER', 0.0, 'NULL'), char
    else:
        return ('ERRTOKEN', '', 0.0, 'NULL'), char